src/messenger.ts

```diff
@@ -0,0 +1,531 @@
+ *
+ * @param type The `type` of message to register against
+ * @param callback The listener callback to register that will receive messages of the given type
+ * @param options Options for the target window
+ */
+export const register: RegisterListener = (type, callback, options?): void => {
+	if (REGISTERED_LISTENERS === 0) {
+		on(options?.window ?? window);
+	}
+
+	const listeners = LISTENERS[type] ?? [];
+
+	if (Array.isArray(listeners) && !listeners.includes(callback)) {
+		LISTENERS[type] = [...listeners, callback];
+		REGISTERED_LISTENERS += 1;
+	}
+};
+
+/**
+ * Register a persistent listener for a given type of iframe message
+ *
+ * @param type The `type` of message to register against
+ * @param callback The persistent listener callback to register that will receive messages of the given type
+ * @param options Options for the target window and whether the callback is persistent
+ */
+export const registerPersistentListener: RegisterPersistentListener = (
+	type,
+	callback,
+	options?,
+): void => {
+	if (REGISTERED_LISTENERS === 0) {
+		on(options?.window ?? window);
+	}
+	LISTENERS[type] = callback;
+	REGISTERED_LISTENERS += 1;
+};
+
+/**
+ * Unregister a callback for a given type
+ *
+ * @param type The type of message to unregister against. An iframe will send
+ * messages annotated with the type
+ * @param callback Optionally include the original callback. If this is included
+ * for a persistent callback this function will be unregistered. If it's
+ * included for a non-persistent callback only the matching callback is removed,
+ * otherwise all callbacks for that type will be unregistered
+ * @param options Option for the target window
+ */
+export const unregister: UnregisterListener = (type, callback, options) => {
+	const listeners = LISTENERS[type];
+
+	if (listeners === undefined) {
+		throw new Error(formatError(error405, type).message);
+	} else if (listeners === callback) {
+		LISTENERS[type] = undefined;
+		REGISTERED_LISTENERS -= 1;
+	} else if (Array.isArray(listeners)) {
+		if (callback === undefined) {
+			LISTENERS[type] = [];
+			REGISTERED_LISTENERS -= listeners.length;
+		} else {
+			LISTENERS[type] = listeners.filter((cb) => {
+				const callbacksEqual = cb === callback;
+				if (callbacksEqual) {
+					REGISTERED_LISTENERS -= 1;
+				}
+				return !callbacksEqual;
+			});
+		}
+	}
+
+	if (REGISTERED_LISTENERS === 0) {
+		off(options?.window ?? window);
+	}
+};
+
+/**
+ * Initialize an array of listener callbacks in a batch
+ *
+ * @param listeners The listener registration functions
+ * @param persistentListeners The persistent listener registration functions
+ */
+export const init = (
+	listeners: Array<
+		(
+			register: RegisterListener,
+			errorHandler: MessengerErrorHandler,
+		) => void
+	>,
+	persistentListeners: Array<
+		(
+			register: RegisterPersistentListener,
+			errorHandler: MessengerErrorHandler,
+		) => void
+	>,
+	errorHandler: MessengerErrorHandler,
```

Comment:
I thought passing in the raven/sentry error handler would be better rather than copying it over, but not totally sure.